core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -98,55 +99,92 @@ object TransactionMarkerChannelManager {
+
+                      case Errors.NOT_COORDINATOR =>
+                        info(s"No longer the coordinator for transactionalId: $transactionalId while trying to append to transaction log, skip writing to transaction log")
+
+                      case Errors.COORDINATOR_NOT_AVAILABLE =>
+                        warn(s"Failed updating transaction state for $transactionalId when appending to transaction log due to ${error.exceptionName}. retrying")
+
+                        // retry appending
+                        txnStateManager.appendTransactionToLog(transactionalId, coordinatorEpoch, newMetadata, retryAppendCallback)
+
+                      case errors: Errors =>
+                        throw new IllegalStateException(s"Unexpected error ${errors.exceptionName} while appending to transaction log for $transactionalId")
+                    }
+
+                  txnStateManager.appendTransactionToLog(transactionalId, coordinatorEpoch, newMetadata, retryAppendCallback)
+                } else {
+                  info(s"Updating $transactionalId's transaction state to $txnMetadata with coordinator epoch $coordinatorEpoch for $transactionalId failed after the transaction markers " +
+                    s"has been sent to brokers. The cached metadata have been changed to $epochAndMetadata since preparing to send markers")
+                }
+              }
+
+            case None =>
+              // this transactional id no longer exists, maybe the corresponding partition has already been migrated out.
+              // we will stop appending the completed log entry to transaction topic as the new leader should be doing it.
+              info(s"Updating $transactionalId's transaction state to $txnMetadata with coordinator epoch $coordinatorEpoch for $transactionalId failed after the transaction message " +
+                s"has been appended to the log. The partition ${txnStateManager.partitionFor(transactionalId)} may have migrated as the metadata is no longer in the cache")
+          }
+
+        case other =>
+          throw new IllegalStateException(s"Unexpected error ${other.exceptionName} before appending to txn log for $transactionalId")
       }
-      transactionMarkerChannel.addRequestToSend(txnTopicPartition,
-        metadataToWrite.pid,
-        metadataToWrite.producerEpoch,
-        result,
-        coordinatorEpoch,
-        metadataToWrite.topicPartitions.toSet)
     }
-  }
 
-  def removeCompleted(txnTopicPartition: Int, pid: Long): Unit = {
-    transactionMarkerChannel.removeCompletedTxn(txnTopicPartition, pid)
+    // watch for both the transactional id and the transaction topic partition id,
+    // so we can cancel all the delayed operations for the same partition id;
+    // NOTE this is only possible because the hashcode of Int / String never overlaps
+    val delayedTxnMarker = new DelayedTxnMarker(txnMetadata, appendToLogCallback)
```

Comment:
Since the timeout for delayedTxnMarker is infinite, do we need a purgatory or just a map?